src/expr/datatype.cpp

```diff
@@ -1306,42 +694,27 @@ Expr DatatypeConstructorArg::getSelector() const {
 Expr DatatypeConstructor::getSelectorInternal( Type domainType, size_t index ) const {
   PrettyCheckArgument(isResolved(), this, "cannot get an internal selector for an unresolved datatype constructor");
   PrettyCheckArgument(index < getNumArgs(), index, "index out of bounds");
-  if( options::dtSharedSelectors() ){
-    computeSharedSelectors( domainType );
-    Assert(d_shared_selectors[domainType].size() == getNumArgs());
-    return d_shared_selectors[domainType][index];
-  }else{
-    return d_args[index].getSelector();
-  }
+  ExprManagerScope ems(d_constructor);
+  TypeNode dtn = TypeNode::fromType(domainType);
+  return d_internal->getSelectorInternal(dtn, index).toExpr();
 }
 
 int DatatypeConstructor::getSelectorIndexInternal( Expr sel ) const {
   PrettyCheckArgument(isResolved(), this, "cannot get an internal selector index for an unresolved datatype constructor");
-  if( options::dtSharedSelectors() ){
-    Assert(sel.getType().isSelector());
-    Type domainType = ((SelectorType)sel.getType()).getDomain();
-    computeSharedSelectors( domainType );
-    std::map< Expr, unsigned >::iterator its = d_shared_selector_index[domainType].find( sel );
-    if( its!=d_shared_selector_index[domainType].end() ){
-      return (int)its->second;
-    }
-  }else{
-    unsigned sindex = Datatype::indexOf(sel);
-    if( getNumArgs() > sindex && d_args[sindex].getSelector() == sel ){
-      return (int)sindex;
-    }
-  }
-  return -1;
+  ExprManagerScope ems(d_constructor);
+  Node seln = Node::fromExpr(sel);
+  return d_internal->getSelectorIndexInternal(seln);
 }
 
 Expr DatatypeConstructorArg::getConstructor() const {
   PrettyCheckArgument(isResolved(), this,
                 "cannot get a associated constructor for argument of an unresolved datatype constructor");
-  return d_constructor;
+  ExprManagerScope ems(d_selector);
```

Comment:
In the other methods, we seem to use `d_constructor` for the `ExprManagerScope`. I'd guess it doesn't matter that we get it from `d_selector` here but maybe we should change it to `d_constructor` for consistency?

Reply:
We use `d_constructor` for `DatatypeConstructor` and `d_selector` for `DatatypeConstructorArg`.

Reply:
Ah yes, I managed to confuse myself, sorry.